onnxruntime/contrib_ops/cpu/transformers/beam_search.cc

```diff
@@ -77,91 +83,144 @@ struct BeamSearchState {
-    // This ensures that the beams in the same group don't produce same tokens every time.
-    for (int i = 0; i < batch_size; i++) {
-      for (int j = 1; j < num_beams; j++) {
-        beam_scores[i * num_beams + j] = -1e9;
-      }
-    }
 
     size_t next_token_size = SafeInt<size_t>(batch_beam_size) * vocab_size;
-    next_token_logits = AllocateBuffer<T>(allocator, next_token_logits_buffer_, next_token_size, true, static_cast<T>(0));
-    next_token_scores = AllocateBuffer<T>(allocator, next_token_scores_buffer_, next_token_size, true, static_cast<T>(0));
+    this->next_token_logits = AllocateBuffer<T>(allocator, next_token_logits_buffer_, next_token_size);
+    this->next_token_scores = AllocateBuffer<float>(allocator, next_token_scores_buffer_, next_token_size);
 
-    next_tokens = AllocateBuffer<int64_t>(allocator, next_tokens_buffer_, SafeInt<size_t>(2) * batch_beam_size, true, static_cast<int64_t>(0));
+    this->next_tokens = AllocateBuffer<int32_t>(allocator, next_tokens_buffer_, SafeInt<size_t>(2) * batch_beam_size);
 
-    next_indices = AllocateBuffer<int64_t>(allocator, next_indices_buffer_, SafeInt<size_t>(2) * batch_beam_size, true, static_cast<int64_t>(0));
+    this->next_indices = AllocateBuffer<int32_t>(allocator, next_indices_buffer_, SafeInt<size_t>(2) * batch_beam_size);
 
-    next_positions = AllocateBuffer<int64_t>(allocator, next_positions_buffer_, batch_beam_size, true, static_cast<int64_t>(0));
+    this->next_positions = AllocateBuffer<int32_t>(allocator, next_positions_buffer_, batch_beam_size);
+
+    this->beam_scores = AllocateBuffer<float>(allocator, beam_scores_buffer_, batch_beam_size);
 
     if (output_scores) {
       size_t elements = SafeInt<size_t>(max_length - sequence_length) * batch_size * num_beams * vocab_size;
-      scores = AllocateBuffer<T>(allocator, scores_buffer_, elements);
-      remaining_scores = scores;
+      this->scores = AllocateBuffer<float>(allocator, scores_buffer_, elements);
+      this->remaining_scores = this->scores;
     }
-
-    // sequences will be initialized later since it has dependency on input_ids
   }
 
  private:
-  BufferUniquePtr beam_scores_buffer_;
   BufferUniquePtr next_token_logits_buffer_;
   BufferUniquePtr next_token_scores_buffer_;
   BufferUniquePtr next_tokens_buffer_;
   BufferUniquePtr next_indices_buffer_;
   BufferUniquePtr next_positions_buffer_;
+  BufferUniquePtr beam_scores_buffer_;
   BufferUniquePtr scores_buffer_;
 };
 
+struct BeamSearchCpuState : public IBeamSearchCpuState {
+  Sequences sequences;
+
+  void Init(AllocatorPtr allocator, size_t batch_beam_size, int max_length, bool is_cuda) {
+    this->sequence_lengths = AllocateBuffer<int32_t>(allocator, sequence_lengths_buffer_, batch_beam_size);
+    this->sequences_space = AllocateBuffer<int32_t>(allocator, sequences_space_buffer_, SafeInt<size_t>(2) * batch_beam_size * max_length);
+
+    if (is_cuda) {
```

Comment:
Is it a bit confused that CUDA buffers occur in a class named BeamSearchCpuState?

Reply:
BeamSearchCpuState has buffers allocated in CPU. Current CUDA implementation need copy some tensors to CPU for some processing.